mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraModSeqProvider.java

```diff
@@ -125,81 +114,76 @@ private PreparedStatement prepareSelect(Session session) {
     }
 
-    private CompletableFuture<Optional<ModSeq>> tryInsertModSeq(CassandraId mailboxId, ModSeq modSeq) {
+    private Mono<ModSeq> tryInsertModSeq(CassandraId mailboxId, ModSeq modSeq) {
         ModSeq nextModSeq = modSeq.next();
         return cassandraAsyncExecutor.executeReturnApplied(
             insert.bind()
                 .setUUID(MAILBOX_ID, mailboxId.asUuid())
                 .setLong(NEXT_MODSEQ, nextModSeq.getValue()))
-            .thenApply(success -> successToModSeq(nextModSeq, success));
+            .flatMap(success -> successToModSeq(nextModSeq, success));
     }
 
-    private CompletableFuture<Optional<ModSeq>> tryUpdateModSeq(CassandraId mailboxId, ModSeq modSeq) {
+    private Mono<ModSeq> tryUpdateModSeq(CassandraId mailboxId, ModSeq modSeq) {
         ModSeq nextModSeq = modSeq.next();
         return cassandraAsyncExecutor.executeReturnApplied(
             update.bind()
                 .setUUID(MAILBOX_ID, mailboxId.asUuid())
                 .setLong(NEXT_MODSEQ, nextModSeq.getValue())
                 .setLong(MOD_SEQ_CONDITION, modSeq.getValue()))
-            .thenApply(success -> successToModSeq(nextModSeq, success));
+            .flatMap(success -> successToModSeq(nextModSeq, success));
     }
 
-    private Optional<ModSeq> successToModSeq(ModSeq modSeq, Boolean success) {
+    private Mono<ModSeq> successToModSeq(ModSeq modSeq, Boolean success) {
         if (success) {
-            return Optional.of(modSeq);
+            return Mono.just(modSeq);
         }
-        return Optional.empty();
+        return Mono.empty();
     }
-    
-    public CompletableFuture<Optional<Long>> nextModSeq(CassandraId mailboxId) {
+
+    public Mono<Long> nextModSeq(CassandraId mailboxId) {
         return findHighestModSeq(mailboxId)
-            .thenCompose(modSeq -> {
+            .flatMap(modSeq -> {
                 if (modSeq.isFirst()) {
                     return tryInsertModSeq(mailboxId, FIRST_MODSEQ);
                 }
                 return tryUpdateModSeq(mailboxId, modSeq);
-            }).thenCompose(firstInsert -> {
-                    if (firstInsert.isPresent()) {
-                        return CompletableFuture.completedFuture(firstInsert);
-                    }
-                    return handleRetries(mailboxId);
-                })
-            .thenApply(optional -> optional.map(ModSeq::getValue));
-    }
-
-    private CompletableFuture<Optional<ModSeq>> handleRetries(CassandraId mailboxId) {
-        return runner.executeAsyncAndRetrieveObject(
-            () -> findHighestModSeq(mailboxId)
-                .thenCompose(newModSeq -> tryUpdateModSeq(mailboxId, newModSeq)));
+            }).switchIfEmpty(handleRetries(mailboxId))
+            .map(ModSeq::getValue);
+    }
+
+    private Mono<ModSeq> handleRetries(CassandraId mailboxId) {
+        return findHighestModSeq(mailboxId)
+                .flatMap(newModSeq -> tryUpdateModSeq(mailboxId, newModSeq));
```

Comment:
Euh does that means I can wait indefinitly without never hitting a max retry?

Reply:
I mean, does this class still tries things a bounded number of times like before (& where?)

FunctionRunnerWithRetry for sure have a max retry bound...